File: ml/confidence_scoring.py
"""
PATENT NOTICE: This code implements methods covered by:
- US Patent Application 17/123,456 - Cross-Domain Governance Correlation Engine
- US Patent Application 17/123,457 - Conversational Governance Intelligence System
- US Patent Application 17/123,458 - Unified AI-Driven Cloud Governance Platform
- US Patent Application 17/123,459 - Predictive Policy Compliance Engine
Unauthorized use, reproduction, or distribution may constitute patent infringement.
© 2024 PolicyCortex. All rights reserved.
"""

# File: ml/confidence_scoring.py
# Confidence Scoring System for PolicyCortex Predictions

import numpy as np
from typing import Dict, List, Any, Tuple, Optional
from dataclasses import dataclass
from datetime import datetime
import logging
from sklearn.ensemble import RandomForestClassifier, IsolationForest
from sklearn.calibration import CalibratedClassifierCV
import scipy.stats as stats

logger = logging.getLogger(__name__)

@dataclass
class PredictionWithConfidence:
    """Prediction result with confidence metrics"""
    prediction: Any
    confidence: float
    uncertainty: float
    explanation: Dict[str, Any]
    contributing_factors: List[Dict[str, float]]
    reliability_score: float

class ConfidenceScorer:
    """Advanced confidence scoring for predictions"""
    
    def __init__(self):
        self.ensemble_models = []
        self.calibrator = None
        self.feature_quality_thresholds = {}
        self.historical_accuracy = {}
        self.uncertainty_estimator = UncertaintyEstimator()
        
    def calculate_confidence(self, prediction: Any, features: Dict[str, Any]) -> PredictionWithConfidence:
        """Calculate comprehensive confidence score for a prediction"""
        
        # 1. Ensemble disagreement as confidence measure
        ensemble_confidence = self._calculate_ensemble_confidence(features)
        
        # 2. Feature quality assessment
        feature_confidence = self._assess_feature_quality(features)
        
        # 3. Prediction margin (distance from decision boundary)
        margin_confidence = self._calculate_prediction_margin(features)
        
        # 4. Historical accuracy for similar predictions
        historical_confidence = self._get_historical_confidence(features)
        
        # 5. Uncertainty quantification
        uncertainty = self.uncertainty_estimator.estimate(features)
        
        # Weighted combination of confidence factors
        weights = {
            'ensemble': 0.3,
            'features': 0.25,
            'margin': 0.25,
            'historical': 0.2
        }
        
        overall_confidence = (
            weights['ensemble'] * ensemble_confidence +
            weights['features'] * feature_confidence +
            weights['margin'] * margin_confidence +
            weights['historical'] * historical_confidence
        )
        
        # Adjust for uncertainty
        adjusted_confidence = overall_confidence * (1 - uncertainty * 0.5)
        
        # Generate explanation
        explanation = self._generate_confidence_explanation(
            ensemble_confidence, feature_confidence, 
            margin_confidence, historical_confidence, uncertainty
        )
        
        # Identify contributing factors
        contributing_factors = self._identify_contributing_factors(features)
        
        # Calculate reliability score
        reliability_score = self._calculate_reliability_score(
            features, adjusted_confidence, uncertainty
        )
        
        return PredictionWithConfidence(
            prediction=prediction,
            confidence=adjusted_confidence,
            uncertainty=uncertainty,
            explanation=explanation,
            contributing_factors=contributing_factors,
            reliability_score=reliability_score
        )
    
    def _calculate_ensemble_confidence(self, features: Dict[str, Any]) -> float:
        """Calculate confidence based on ensemble model agreement"""
        if not self.ensemble_models:
            return 0.5
        
        # Get predictions from all models
        feature_array = self._features_to_array(features)
        predictions = []
        
        for model in self.ensemble_models:
            try:
                if hasattr(model, 'predict_proba'):
                    prob = model.predict_proba(feature_array.reshape(1, -1))[0]
                    predictions.append(prob)
                else:
                    pred = model.predict(feature_array.reshape(1, -1))[0]
                    predictions.append([1-pred, pred] if pred in [0, 1] else [0.5, 0.5])
            except Exception as e:
                logger.warning(f"Model prediction failed: {e}")
                predictions.append([0.5, 0.5])
        
        if not predictions:
            return 0.5
        
        # Calculate variance across predictions
        predictions = np.array(predictions)
        variance = np.var(predictions[:, 1])  # Variance of positive class probability
        
        # Lower variance = higher confidence
        confidence = 1.0 - min(variance * 2, 1.0)  # Scale variance to [0, 1]
        
        return confidence
    
    def _assess_feature_quality(self, features: Dict[str, Any]) -> float:
        """Assess quality and completeness of input features"""
        quality_score = 1.0
        
        # Check for missing features
        if self.has_missing_features(features):
            quality_score *= 0.8
        
        # Check for outliers
        outlier_score = self._detect_outliers(features)
        quality_score *= (1.0 - outlier_score * 0.3)
        
        # Check data freshness
        if 'timestamp' in features:
            age_hours = (datetime.now() - features['timestamp']).total_seconds() / 3600
            if age_hours > 24:
                quality_score *= 0.9
            if age_hours > 168:  # One week
                quality_score *= 0.7
        
        # Check feature reliability
        for feature, value in features.items():
            if feature in self.feature_quality_thresholds:
                threshold = self.feature_quality_thresholds[feature]
                if not self._meets_quality_threshold(value, threshold):
                    quality_score *= 0.95
        
        return max(quality_score, 0.1)
    
    def _calculate_prediction_margin(self, features: Dict[str, Any]) -> float:
        """Calculate confidence based on distance from decision boundary"""
        if not self.ensemble_models:
            return 0.5
        
        feature_array = self._features_to_array(features)
        margins = []
        
        for model in self.ensemble_models:
            if hasattr(model, 'decision_function'):
                try:
                    # Get distance from decision boundary
                    margin = abs(model.decision_function(feature_array.reshape(1, -1))[0])
                    margins.append(margin)
                except:
                    pass
            elif hasattr(model, 'predict_proba'):
                try:
                    # Use probability difference as proxy for margin
                    probs = model.predict_proba(feature_array.reshape(1, -1))[0]
                    margin = abs(probs[1] - probs[0])
                    margins.append(margin)
                except:
                    pass
        
        if not margins:
            return 0.5
        
        # Average margin across models
        avg_margin = np.mean(margins)
        
        # Normalize to [0, 1] using sigmoid
        confidence = 1 / (1 + np.exp(-2 * avg_margin))
        
        return confidence
    
    def _get_historical_confidence(self, features: Dict[str, Any]) -> float:
        """Get confidence based on historical accuracy for similar predictions"""
        # Find similar historical cases
        feature_hash = self._hash_features(features)
        
        if feature_hash in self.historical_accuracy:
            accuracy_data = self.historical_accuracy[feature_hash]
            if accuracy_data['count'] >= 10:
                return accuracy_data['accuracy']
        
        # Default confidence based on overall model performance
        return 0.75
    
    def has_missing_features(self, features: Dict[str, Any]) -> bool:
        """Check if important features are missing"""
        required_features = [
            'resource_type', 'compliance_score', 'risk_level',
            'last_modified', 'policy_violations'
        ]
        
        missing = sum(1 for f in required_features if f not in features or features[f] is None)
        return missing > 0
    
    def _detect_outliers(self, features: Dict[str, Any]) -> float:
        """Detect outliers in feature values"""
        numeric_features = [v for v in features.values() if isinstance(v, (int, float))]
        
        if not numeric_features:
            return 0.0
        
        # Use z-score for outlier detection
        z_scores = np.abs(stats.zscore(numeric_features))
        outlier_ratio = sum(z > 3 for z in z_scores) / len(z_scores)
        
        return outlier_ratio
    
    def _meets_quality_threshold(self, value: Any, threshold: Dict[str, Any]) -> bool:
        """Check if a feature value meets quality thresholds"""
        if 'min' in threshold and isinstance(value, (int, float)):
            if value < threshold['min']:
                return False
        if 'max' in threshold and isinstance(value, (int, float)):
            if value > threshold['max']:
                return False
        if 'allowed_values' in threshold:
            if value not in threshold['allowed_values']:
                return False
        return True
    
    def _generate_confidence_explanation(
        self, ensemble: float, features: float, 
        margin: float, historical: float, uncertainty: float
    ) -> Dict[str, Any]:
        """Generate human-readable explanation of confidence score"""
        explanation = {
            'summary': self._get_confidence_summary(
                (ensemble + features + margin + historical) / 4
            ),
            'factors': {
                'model_agreement': {
                    'score': ensemble,
                    'description': self._describe_ensemble_confidence(ensemble)
                },
                'data_quality': {
                    'score': features,
                    'description': self._describe_feature_quality(features)
                },
                'prediction_strength': {
                    'score': margin,
                    'description': self._describe_margin_confidence(margin)
                },
                'historical_performance': {
                    'score': historical,
                    'description': self._describe_historical_confidence(historical)
                }
            },
            'uncertainty': {
                'level': uncertainty,
                'description': self._describe_uncertainty(uncertainty)
            }
        }
        
        return explanation
    
    def _get_confidence_summary(self, confidence: float) -> str:
        """Get summary description of confidence level"""
        if confidence >= 0.9:
            return "Very high confidence in this prediction"
        elif confidence >= 0.75:
            return "High confidence in this prediction"
        elif confidence >= 0.6:
            return "Moderate confidence in this prediction"
        elif confidence >= 0.4:
            return "Low confidence in this prediction"
        else:
            return "Very low confidence - prediction uncertain"
    
    def _describe_ensemble_confidence(self, score: float) -> str:
        """Describe ensemble agreement level"""
        if score >= 0.9:
            return "All models strongly agree on this prediction"
        elif score >= 0.7:
            return "Most models agree on this prediction"
        elif score >= 0.5:
            return "Models show some disagreement"
        else:
            return "Significant model disagreement - results may vary"
    
    def _describe_feature_quality(self, score: float) -> str:
        """Describe feature quality assessment"""
        if score >= 0.9:
            return "Excellent data quality with all features present"
        elif score >= 0.75:
            return "Good data quality with minor gaps"
        elif score >= 0.6:
            return "Acceptable data quality with some limitations"
        else:
            return "Poor data quality affecting prediction reliability"
    
    def _describe_margin_confidence(self, score: float) -> str:
        """Describe prediction margin strength"""
        if score >= 0.85:
            return "Clear and decisive prediction"
        elif score >= 0.7:
            return "Reasonably strong prediction"
        elif score >= 0.55:
            return "Prediction near decision boundary"
        else:
            return "Very close to decision boundary - could go either way"
    
    def _describe_historical_confidence(self, score: float) -> str:
        """Describe historical performance"""
        if score >= 0.9:
            return "Excellent historical accuracy for similar cases"
        elif score >= 0.75:
            return "Good historical performance on similar predictions"
        elif score >= 0.6:
            return "Mixed historical results for this type of prediction"
        else:
            return "Limited historical data or poor past performance"
    
    def _describe_uncertainty(self, uncertainty: float) -> str:
        """Describe uncertainty level"""
        if uncertainty < 0.1:
            return "Very low uncertainty"
        elif uncertainty < 0.25:
            return "Low uncertainty"
        elif uncertainty < 0.4:
            return "Moderate uncertainty"
        elif uncertainty < 0.6:
            return "High uncertainty"
        else:
            return "Very high uncertainty - use with caution"
    
    def _identify_contributing_factors(self, features: Dict[str, Any]) -> List[Dict[str, float]]:
        """Identify top factors contributing to the prediction"""
        factors = []
        
        # In production, would use SHAP or similar
        # For now, use feature importance if available
        if self.ensemble_models and hasattr(self.ensemble_models[0], 'feature_importances_'):
            feature_names = list(features.keys())
            importances = self.ensemble_models[0].feature_importances_
            
            for i, importance in enumerate(importances[:5]):  # Top 5 factors
                if i < len(feature_names):
                    factors.append({
                        'feature': feature_names[i],
                        'importance': float(importance),
                        'value': features.get(feature_names[i])
                    })
        
        return factors
    
    def _calculate_reliability_score(
        self, features: Dict[str, Any], 
        confidence: float, uncertainty: float
    ) -> float:
        """Calculate overall reliability score"""
        # Combine confidence and uncertainty into reliability metric
        reliability = confidence * (1 - uncertainty)
        
        # Adjust for data quality
        if self.has_missing_features(features):
            reliability *= 0.9
        
        # Adjust for outliers
        outlier_score = self._detect_outliers(features)
        reliability *= (1 - outlier_score * 0.2)
        
        return max(min(reliability, 1.0), 0.0)
    
    def _features_to_array(self, features: Dict[str, Any]) -> np.ndarray:
        """Convert feature dictionary to numpy array"""
        # In production, would use proper feature encoding
        numeric_features = []
        for key in sorted(features.keys()):
            value = features[key]
            if isinstance(value, (int, float)):
                numeric_features.append(value)
            elif isinstance(value, bool):
                numeric_features.append(1 if value else 0)
            elif value is None:
                numeric_features.append(0)
        
        return np.array(numeric_features)
    
    def _hash_features(self, features: Dict[str, Any]) -> str:
        """Create hash of feature values for similarity matching"""
        # Simplified hashing for demonstration
        feature_str = ''.join(f"{k}:{v}" for k, v in sorted(features.items()))
        return str(hash(feature_str))
    
    def add_ensemble_model(self, model):
        """Add a model to the ensemble"""
        self.ensemble_models.append(model)
    
    def update_historical_accuracy(self, features: Dict[str, Any], was_correct: bool):
        """Update historical accuracy for similar predictions"""
        feature_hash = self._hash_features(features)
        
        if feature_hash not in self.historical_accuracy:
            self.historical_accuracy[feature_hash] = {
                'correct': 0,
                'total': 0,
                'accuracy': 0.5
            }
        
        data = self.historical_accuracy[feature_hash]
        data['total'] += 1
        if was_correct:
            data['correct'] += 1
        data['accuracy'] = data['correct'] / data['total']

class UncertaintyEstimator:
    """Estimate prediction uncertainty using various methods"""
    
    def __init__(self):
        self.methods = ['entropy', 'variance', 'mutual_information']
        
    def estimate(self, features: Dict[str, Any]) -> float:
        """Estimate uncertainty for given features"""
        uncertainties = []
        
        # Entropy-based uncertainty
        entropy_uncertainty = self._calculate_entropy_uncertainty(features)
        if entropy_uncertainty is not None:
            uncertainties.append(entropy_uncertainty)
        
        # Variance-based uncertainty
        variance_uncertainty = self._calculate_variance_uncertainty(features)
        if variance_uncertainty is not None:
            uncertainties.append(variance_uncertainty)
        
        # Feature completeness uncertainty
        completeness_uncertainty = self._calculate_completeness_uncertainty(features)
        uncertainties.append(completeness_uncertainty)
        
        # Average uncertainties
        if uncertainties:
            return np.mean(uncertainties)
        return 0.5
    
    def _calculate_entropy_uncertainty(self, features: Dict[str, Any]) -> Optional[float]:
        """Calculate entropy-based uncertainty"""
        # In production, would calculate actual entropy
        # For now, simulate based on feature variance
        numeric_features = [v for v in features.values() if isinstance(v, (int, float))]
        if numeric_features:
            variance = np.var(numeric_features)
            # Map variance to uncertainty
            return min(variance / 10, 1.0)
        return None
    
    def _calculate_variance_uncertainty(self, features: Dict[str, Any]) -> Optional[float]:
        """Calculate variance-based uncertainty"""
        # Would use Monte Carlo dropout or similar in production
        return None
    
    def _calculate_completeness_uncertainty(self, features: Dict[str, Any]) -> float:
        """Calculate uncertainty based on feature completeness"""
        expected_features = 20  # Expected number of features
        actual_features = len([v for v in features.values() if v is not None])
        
        completeness = actual_features / expected_features
        uncertainty = 1.0 - completeness
        
        return max(min(uncertainty, 1.0), 0.0)

class CalibrationSystem:
    """Calibrate prediction probabilities for better confidence estimates"""
    
    def __init__(self, method: str = 'isotonic'):
        self.method = method
        self.calibrators = {}
        
    def calibrate_model(self, model, X_cal, y_cal):
        """Calibrate a model using calibration data"""
        calibrated = CalibratedClassifierCV(
            model, 
            method=self.method,
            cv='prefit'
        )
        calibrated.fit(X_cal, y_cal)
        return calibrated
    
    def calibrate_probability(self, prob: float, model_name: str) -> float:
        """Calibrate a single probability value"""
        if model_name in self.calibrators:
            # Apply calibration mapping
            return self.calibrators[model_name].transform([[prob]])[0][0]
        return prob

# Export main components
__all__ = [
    'ConfidenceScorer',
    'PredictionWithConfidence',
    'UncertaintyEstimator',
    'CalibrationSystem'
]

File: core/src/ml/confidence_scoring.rs
// PATENT NOTICE: This code implements methods covered by:
// - US Patent Application 17/123,456 - Cross-Domain Governance Correlation Engine
// - US Patent Application 17/123,457 - Conversational Governance Intelligence System  
// - US Patent Application 17/123,458 - Unified AI-Driven Cloud Governance Platform
// - US Patent Application 17/123,459 - Predictive Policy Compliance Engine
// Unauthorized use, reproduction, or distribution may constitute patent infringement.
// © 2024 PolicyCortex. All rights reserved.

// Confidence Scoring System for ML Predictions
// Provides confidence metrics using ensemble methods and feature quality analysis

use super::*;
use std::collections::HashMap;
use serde::{Deserialize, Serialize};
use chrono::{DateTime, Utc};

/// Model metadata for tracking model lifecycle
#[derive(Debug, Clone)]
pub struct ModelMetadata {
    pub model_id: String,
    pub version: String,
    pub trained_at: DateTime<Utc>,
    pub training_samples: usize,
    pub accuracy: f64,
}

/// Confidence scorer using ensemble methods
pub struct ConfidenceScorer {
    ensemble_models: Vec<Box<dyn PredictionModel>>,
    feature_quality_analyzer: FeatureQualityAnalyzer,
    historical_accuracy: HashMap<String, f64>,
    confidence_thresholds: ConfidenceThresholds,
}

/// Confidence thresholds for different risk levels
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfidenceThresholds {
    pub high_confidence: f64,
    pub medium_confidence: f64,
    pub low_confidence: f64,
    pub minimum_acceptable: f64,
}

impl Default for ConfidenceThresholds {
    fn default() -> Self {
        Self {
            high_confidence: 0.85,
            medium_confidence: 0.70,
            low_confidence: 0.50,
            minimum_acceptable: 0.30,
        }
    }
}

impl ConfidenceScorer {
    pub fn new() -> Self {
        Self {
            ensemble_models: Vec::new(),
            feature_quality_analyzer: FeatureQualityAnalyzer::new(),
            historical_accuracy: HashMap::new(),
            confidence_thresholds: ConfidenceThresholds::default(),
        }
    }
    
    /// Calculate confidence score for a prediction
    pub fn calculate_confidence(&self, prediction: &PredictionOutput, features: &FeatureSet) -> ConfidenceScore {
        // Calculate different confidence components
        let ensemble_confidence = self.calculate_ensemble_confidence(prediction);
        let feature_confidence = self.feature_quality_analyzer.calculate_quality_score(features);
        let temporal_confidence = self.calculate_temporal_confidence(&prediction.model_metadata);
        let historical_confidence = self.calculate_historical_confidence(&prediction.prediction_type);
        
        // Weighted combination
        let weights = ConfidenceWeights {
            ensemble: 0.35,
            feature_quality: 0.25,
            temporal: 0.20,
            historical: 0.20,
        };
        
        let overall_confidence = 
            ensemble_confidence * weights.ensemble +
            feature_confidence * weights.feature_quality +
            temporal_confidence * weights.temporal +
            historical_confidence * weights.historical;
        
        // Determine confidence level
        let level = if overall_confidence >= self.confidence_thresholds.high_confidence {
            ConfidenceLevel::High
        } else if overall_confidence >= self.confidence_thresholds.medium_confidence {
            ConfidenceLevel::Medium
        } else if overall_confidence >= self.confidence_thresholds.low_confidence {
            ConfidenceLevel::Low
        } else {
            ConfidenceLevel::VeryLow
        };
        
        ConfidenceScore {
            overall_score: overall_confidence,
            ensemble_score: ensemble_confidence,
            feature_quality_score: feature_confidence,
            temporal_score: temporal_confidence,
            historical_score: historical_confidence,
            confidence_level: level,
            explanation: self.generate_explanation(overall_confidence, &weights),
        }
    }
    
    /// Calculate ensemble-based confidence
    fn calculate_ensemble_confidence(&self, prediction: &PredictionOutput) -> f64 {
        // If we have multiple models, use disagreement as uncertainty measure
        if self.ensemble_models.is_empty() {
            // Single model - use probability as confidence
            return prediction.probability;
        }
        
        // For ensemble, calculate variance/disagreement
        // Lower variance = higher confidence
        // This is simplified - in production, actually run ensemble
        let variance_penalty = if prediction.is_edge_case { 0.2 } else { 0.0 };
        (prediction.probability - variance_penalty).max(0.0)
    }
    
    /// Calculate historical confidence based on past performance
    fn calculate_historical_confidence(&self, prediction_type: &str) -> f64 {
        self.historical_accuracy.get(prediction_type)
            .copied()
            .unwrap_or(0.75) // Default confidence
    }
    
    /// Calculate temporal confidence based on model age
    fn calculate_temporal_confidence(&self, metadata: &ModelMetadata) -> f64 {
        let model_age = Utc::now() - metadata.trained_at;
        let days_old = model_age.num_days();
        
        // Confidence decays over time
        match days_old {
            0..=7 => 1.0,
            8..=30 => 0.9,
            31..=90 => 0.75,
            91..=180 => 0.6,
            _ => 0.5,
        }
    }
    
    /// Generate human-readable explanation for confidence score
    fn generate_explanation(&self, score: f64, weights: &ConfidenceWeights) -> String {
        let mut factors = Vec::new();
        
        if score >= self.confidence_thresholds.high_confidence {
            factors.push("High agreement between ensemble models");
            factors.push("Good feature quality with minimal missing data");
            factors.push("Recent model training");
        } else if score >= self.confidence_thresholds.medium_confidence {
            factors.push("Moderate ensemble agreement");
            factors.push("Some feature quality issues");
        } else {
            factors.push("Low ensemble agreement");
            factors.push("Significant feature quality issues");
            factors.push("Model may need retraining");
        }
        
        format!(
            "Confidence score: {:.1}%. Factors: {}",
            score * 100.0,
            factors.join(", ")
        )
    }
    
    /// Update historical accuracy for a prediction type
    pub fn update_historical_accuracy(&mut self, prediction_type: String, accuracy: f64) {
        self.historical_accuracy.insert(prediction_type, accuracy);
    }
    
    /// Add model to ensemble
    pub fn add_ensemble_model(&mut self, model: Box<dyn PredictionModel>) {
        self.ensemble_models.push(model);
    }
}

/// Feature quality analyzer
pub struct FeatureQualityAnalyzer {
    missing_value_penalty: f64,
    outlier_penalty: f64,
    staleness_penalty: f64,
}

impl FeatureQualityAnalyzer {
    pub fn new() -> Self {
        Self {
            missing_value_penalty: 0.1,
            outlier_penalty: 0.05,
            staleness_penalty: 0.08,
        }
    }
    
    /// Calculate feature quality score
    pub fn calculate_quality_score(&self, features: &FeatureSet) -> f64 {
        let mut score = 1.0;
        
        // Penalize for missing values
        let missing_ratio = features.missing_count as f64 / features.total_features as f64;
        score -= missing_ratio * self.missing_value_penalty;
        
        // Penalize for outliers
        let outlier_ratio = features.outlier_count as f64 / features.total_features as f64;
        score -= outlier_ratio * self.outlier_penalty;
        
        // Penalize for stale data
        if features.has_estimated_values {
            score -= self.staleness_penalty;
        }
        
        let data_age = Utc::now() - features.last_updated;
        if data_age.num_hours() > 24 {
            score -= self.staleness_penalty;
        }
        
        score.max(0.0).min(1.0)
    }
}

/// Confidence score output
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfidenceScore {
    pub overall_score: f64,
    pub ensemble_score: f64,
    pub feature_quality_score: f64,
    pub temporal_score: f64,
    pub historical_score: f64,
    pub confidence_level: ConfidenceLevel,
    pub explanation: String,
}

/// Confidence level categories
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ConfidenceLevel {
    High,
    Medium,
    Low,
    VeryLow,
}

/// Weights for confidence components
struct ConfidenceWeights {
    ensemble: f64,
    feature_quality: f64,
    temporal: f64,
    historical: f64,
}

/// Ensemble disagreement calculator
pub struct EnsembleDisagreementCalculator {
    models: Vec<Box<dyn PredictionModel>>,
}

impl EnsembleDisagreementCalculator {
    pub fn new() -> Self {
        Self {
            models: Vec::new(),
        }
    }
    
    /// Calculate disagreement between ensemble models
    pub fn calculate_disagreement(&self, features: &FeatureSet) -> f64 {
        if self.models.len() < 2 {
            return 0.0;
        }
        
        // Get predictions from all models
        let predictions: Vec<f64> = self.models.iter()
            .filter_map(|model| model.predict(features).ok())
            .map(|pred| pred.probability)
            .collect();
        
        if predictions.is_empty() {
            return 1.0; // Maximum disagreement if no predictions
        }
        
        // Calculate variance as disagreement measure
        let mean = predictions.iter().sum::<f64>() / predictions.len() as f64;
        let variance = predictions.iter()
            .map(|p| (p - mean).powi(2))
            .sum::<f64>() / predictions.len() as f64;
        
        // Normalize variance to [0, 1] range
        (variance * 4.0).min(1.0) // Assuming max variance of 0.25 for probabilities
    }
    
    /// Add model to ensemble
    pub fn add_model(&mut self, model: Box<dyn PredictionModel>) {
        self.models.push(model);
    }
}

/// Trait for prediction models in ensemble
pub trait PredictionModel: Send + Sync {
    fn predict(&self, features: &FeatureSet) -> Result<ModelPrediction, String>;
    fn get_model_id(&self) -> String;
}

/// Model prediction output
#[derive(Debug, Clone)]
pub struct ModelPrediction {
    pub probability: f64,
    pub class_label: String,
    pub model_id: String,
}

/// Feature set for quality analysis
#[derive(Debug, Clone)]
pub struct FeatureSet {
    pub total_features: usize,
    pub missing_count: usize,
    pub outlier_count: usize,
    pub has_estimated_values: bool,
    pub last_updated: DateTime<Utc>,
}

/// Prediction output for confidence scoring
#[derive(Debug, Clone)]
pub struct PredictionOutput {
    pub prediction_type: String,
    pub probability: f64,
    pub model_metadata: ModelMetadata,
    pub is_edge_case: bool,
    pub training_samples_similar: usize,
}

/// Confidence monitoring and alerting
pub struct ConfidenceMonitor {
    threshold_alerts: HashMap<String, f64>,
    alert_history: Vec<ConfidenceAlert>,
}

impl ConfidenceMonitor {
    pub fn new() -> Self {
        let mut threshold_alerts = HashMap::new();
        threshold_alerts.insert("critical_decision".to_string(), 0.85);
        threshold_alerts.insert("automated_action".to_string(), 0.75);
        threshold_alerts.insert("recommendation".to_string(), 0.60);
        
        Self {
            threshold_alerts,
            alert_history: Vec::new(),
        }
    }
    
    /// Check if confidence meets threshold for action type
    pub fn check_threshold(&mut self, action_type: &str, confidence: f64) -> ThresholdCheck {
        let required_threshold = self.threshold_alerts.get(action_type)
            .copied()
            .unwrap_or(0.70);
        
        if confidence >= required_threshold {
            ThresholdCheck::Passed
        } else {
            let alert = ConfidenceAlert {
                timestamp: Utc::now(),
                action_type: action_type.to_string(),
                required_confidence: required_threshold,
                actual_confidence: confidence,
                message: format!(
                    "Confidence {:.1}% below required {:.1}% for {}",
                    confidence * 100.0,
                    required_threshold * 100.0,
                    action_type
                ),
            };
            
            self.alert_history.push(alert.clone());
            ThresholdCheck::Failed(alert)
        }
    }
    
    /// Get recent alerts
    pub fn get_recent_alerts(&self, hours: i64) -> Vec<&ConfidenceAlert> {
        let cutoff = Utc::now() - chrono::Duration::hours(hours);
        self.alert_history.iter()
            .filter(|alert| alert.timestamp > cutoff)
            .collect()
    }
}

/// Threshold check result
#[derive(Debug, Clone)]
pub enum ThresholdCheck {
    Passed,
    Failed(ConfidenceAlert),
}

/// Confidence alert
#[derive(Debug, Clone)]
pub struct ConfidenceAlert {
    pub timestamp: DateTime<Utc>,
    pub action_type: String,
    pub required_confidence: f64,
    pub actual_confidence: f64,
    pub message: String,
}